src/vcpkg/base/downloads.cpp

```diff
@@ -447,31 +450,58 @@ namespace vcpkg::Downloads
             const auto dir = download_path_part_path.parent_path();
             fs.create_directories(dir, VCPKG_LINE_INFO);
 
-            auto f = fs.open_for_write(download_path_part_path, VCPKG_LINE_INFO);
-
             const auto sanitized_url = replace_secrets(url, secrets);
             Debug::print("Downloading ", sanitized_url, "\n");
             static auto s = WinHttpSession::make().value_or_exit(VCPKG_LINE_INFO);
-            auto conn = WinHttpConnection::make(s.m_hSession.get(), hostname, port);
-            if (!conn)
+            for (size_t trials = 0; trials < 4; ++trials)
             {
-                Strings::append(errors, sanitized_url, ": ", conn.error(), '\n');
-                return false;
-            }
-            auto req = WinHttpRequest::make(
-                conn.get()->m_hConnect.get(), split_uri.path_query_fragment, split_uri.scheme == "https");
-            if (!req)
-            {
-                Strings::append(errors, sanitized_url, ": ", req.error(), '\n');
-                return false;
-            }
-            auto forall_data = req.get()->forall_data([&f](Span<char> span) { f.write(span.data(), 1, span.size()); });
-            if (!forall_data)
-            {
-                Strings::append(errors, sanitized_url, ": ", forall_data.error(), '\n');
-                return false;
+                if (trials > 0)
+                {
+                    // 0.5s, 1s, 2s
+                    Debug::print("Download failed -- retrying after ", 250 << trials, " ms.\n");
+                    std::this_thread::sleep_for(std::chrono::milliseconds(250 << trials));
+                }
+                auto conn = WinHttpConnection::make(s.m_hSession.get(), hostname, port);
+                if (!conn)
+                {
+                    Strings::append(errors, sanitized_url, ": ", conn.error(), '\n');
+                    continue;
+                }
+                auto req = WinHttpRequest::make(
+                    conn.get()->m_hConnect.get(), split_uri.path_query_fragment, split_uri.scheme == "https");
+                if (!req)
+                {
+                    Strings::append(errors, sanitized_url, ": ", req.error(), '\n');
+                    continue;
+                }
+
+                auto maybe_status = req.get()->query_status();
+                if (auto status = maybe_status.get())
+                {
+                    if (*status < 200 || *status >= 300)
+                    {
+                        Strings::append(errors, sanitized_url, ": failed: status code ", *status, '\n');
+                        return false;
+                    }
+                }
+                else
+                {
+                    Strings::append(errors, sanitized_url, ": ", maybe_status.error(), '\n');
+                    continue;
+                }
+
+                const auto f = fs.open_for_write(download_path_part_path, VCPKG_LINE_INFO);
```

Comment:
Instead of moving this down here I think it would be better to introduce a "clear" to `WriteFileHandle` so that we eagerly create the file while waiting for DNS resolution etc.